src/Workspaces/Core/Portable/SymbolId/SymbolKey.Visitor.cs

```diff
@@ -1,140 +1,407 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+using System;
 using System.Collections.Generic;
+using System.Collections.Immutable;
 using System.Diagnostics;
+using System.Text;
 using System.Threading;
+using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis
 {
-    internal abstract partial class SymbolKey
+    internal partial class SymbolKey
     {
-        private class Visitor : SymbolVisitor<SymbolKey>
+        private enum SymbolKeyType
         {
-            internal readonly Dictionary<ISymbol, SymbolKey> SymbolCache = new Dictionary<ISymbol, SymbolKey>();
-            internal readonly Compilation Compilation;
-            internal readonly CancellationToken CancellationToken;
+            Alias = 'A',
+            BodyLevel = 'B',
+            ConstructedMethod = 'C',
+            NamedType = 'D',
+            ErrorType = 'E',
+            Field = 'F',
+            DynamicType = 'I',
+            Method = 'M',
+            Namespace = 'N',
+            PointerType = 'O',
+            Parameter = 'P',
+            Property = 'Q',
+            ArrayType = 'R',
+            Assembly = 'S',
+            TupleType = 'T',
+            Module = 'U',
+            Event = 'V',
+            ReducedExtensionMethod = 'X',
+            TypeParameter = 'Y',
 
-            public Visitor(Compilation compilation, CancellationToken cancellationToken)
+            // Not to be confused with ArrayType.  This indicates an array of elements in the stream.
+            Array = '%',
+            Reference = '#',
+            Null = '!',
+            TypeParameterOrdinal = '@',
+        }
+
+        private class Visitor : SymbolVisitor<object>, IDisposable
```

Comment:
Its odd now that this is named Visitor, when it is really doing the job of a Serializer.  Visiting is just how it works.